### PR TITLE
Settlement report csv

### PR DESCRIPTION
**Pull request by 00eadeyemi**
### Features 

- Settlement report table is now responsive to screen resizing.
- Settlement report shown in the view is now downloadable as a csv. 

[Delivers #126586279](https://www.pivotaltracker.com/story/show/126586279)

### CODE REVIEW

**Comment by 00eadeyemi** [comment]
@srleach please pre-review

File: resources/views/settlements/settlement_report.blade.php [hunk 1]
```diff
@@ -3,9 +3,10 @@
 @section('content')
 
     <h1>
-        SETTLEMENT REPORT
-        <div class="btn-group pull-right">
-            <a href="{!! Request::url() !!}/?download=csv&amp;limit=5000" class="btn btn-default"><span class="glyphicon glyphicon-save" aria-hidden="true"></span> Download CSV</a>
+        Settlement Report
+        <div class="btn-toolbar pull-right">
+            <a href="{{Request::url()}}/?download=csv&amp;source=view_data&amp;filename=export_view_filename&amp;limit=5000" class="btn btn-default"><span class="glyphicon glyphicon-save"></span> Download Report</a>
```
Comment (srleach) [inline resources/views/settlements/settlement_report.blade.php]: - I wouldn't put all of these params in the URL. Set defaults, and just have `/?download=csv`
- On that note, I would change the download key to format. So you'd have each url as `/?format=csv`
- What is `filename=export_api_filename` ???

File: tests/Controllers/SettlementControllerTest.php [hunk 2]
```diff
@@ -199,4 +201,38 @@ public function testApplySettlementAmounts()
...
+        $method1->setAccessible(TRUE);
+        $method1->invokeArgs($this->settlementControllerObject,[&$this->settlementReport]);
+
+        $method = new ReflectionMethod('App\Http\Controllers\SettlementsController', 'flattenViewReport');
+        $method->setAccessible(TRUE);
+        $flattenSettlementReport = $method->invokeArgs($this->settlementControllerObject,[$this->settlementReport]);
+
+        $this->assertArraySubset($lastRow,$flattenSettlementReport[2]);
+    }
+
```
Comment (srleach) [inline tests/Controllers/SettlementControllerTest.php]: CS

File: tests/Controllers/SettlementControllerTest.php [hunk 3]
```diff
@@ -199,4 +201,38 @@ public function testApplySettlementAmounts()
...
 
     }
+
+    /**
+     * @author EA
+     */
+    public function testFlattenViewReport()
+    {
+        $lastRow =
+            [
```
Comment (srleach) [inline tests/Controllers/SettlementControllerTest.php]: CS

File: tests/Controllers/SettlementControllerTest.php [hunk 4]
```diff
@@ -199,4 +201,38 @@ public function testApplySettlementAmounts()
...
+                'Application ID' => 1000000073,
+                'Retailer Reference' =>"86124bb7-57e6-0d91-075d",
+                'Order Amount' => "600.00",
+                'Type' => "Cancellation",
+                'Deposit' => "-60.00",
+                'Loan Amount' => "-540.00",
+                'Subsidy' => "1.00",
+                'Adjustment' => "0.00",
+                'Settlement Amount' => "-599.00",
+            ];
```
Comment (srleach) [inline tests/Controllers/SettlementControllerTest.php]: CS

File: app/Http/Controllers/SettlementsController.php [hunk 5]
```diff
@@ -118,7 +122,7 @@ public function settlementReport($merchant, $id)
      * @param array $report
      * @return array
      */
-    private function flattenReport(array $report)
+    private function flattenApiReport(array $report)
```
Comment (srleach) [inline app/Http/Controllers/SettlementsController.php]: flattenRawReport

File: app/Http/Controllers/SettlementsController.php [hunk 6]
```diff
@@ -107,9 +107,13 @@ public function settlementReport($merchant, $id)
...
             'settlementReport' => $settlementReport,
             'installation' => Application::where('ext_id', '=', $settlementReport['id'])->first(),
-            'api_data' => $this->flattenReport($settlementReport),
-            'export_custom_filename' => 'settlement-report-'.$settlementReport['id'].'-'
+            'api_data' => $this->flattenApiReport($settlementReport),
+            'export_api_filename' => 'settlement-raw-'.$settlementReport['id'].'-'
                 .date_format(DateTime::createFromFormat('Y-m-d',$settlementReport['settlement_date']),'Ymd'),
+            'view_data' => $this->flattenViewReport($settlementReport),
+            'export_view_filename' => 'settlement-report-'.$settlementReport['id'].'-'
+                .date_format(DateTime::createFromFormat('Y-m-d',$settlementReport['settlement_date']),'Ymd'),
```
Comment (srleach) [inline app/Http/Controllers/SettlementsController.php]: CS

File: app/Http/Controllers/SettlementsController.php [hunk 7]
```diff
@@ -107,9 +107,13 @@ public function settlementReport($merchant, $id)
         return View('settlements.settlement_report', [
             'settlementReport' => $settlementReport,
             'installation' => Application::where('ext_id', '=', $settlementReport['id'])->first(),
-            'api_data' => $this->flattenReport($settlementReport),
-            'export_custom_filename' => 'settlement-report-'.$settlementReport['id'].'-'
+            'api_data' => $this->flattenApiReport($settlementReport),
+            'export_api_filename' => 'settlement-raw-'.$settlementReport['id'].'-'
```
Comment (srleach) [inline app/Http/Controllers/SettlementsController.php]: CS

File: app/Http/Controllers/SettlementsController.php [hunk 8]
```diff
@@ -127,21 +131,49 @@ private function flattenReport(array $report)
                 $rtn[] = [
                     'Order Date' => date('d/m/Y', strtotime($settlement['received_date'])),
                     'Customer' => $settlement['customer_name'],
-                    'Postcode' => $settlement['application_postcode'],
+                    'Post Code' => $settlement['application_postcode'],
                     'Retailer Reference' => $settlement['order_reference'],
-                    'Order Amount £' =>  number_format( $settlement['order_amount']/100, 2),
+                    'Order Amount' =>  number_format( $settlement['order_amount']/100,2,'.',''),
```
Comment (srleach) [inline app/Http/Controllers/SettlementsController.php]: CS

File: app/Http/Controllers/SettlementsController.php [hunk 9]
```diff
@@ -127,21 +131,49 @@ private function flattenReport(array $report)
...
                     'Notification Date' => date('d/m/Y', strtotime($settlement['captured_date'])),
                     'Type' => $tx['type'],
                     'Description' => $settlement['description'],
-                    'Deposit £' =>  number_format( $settlement['deposit']/100, 2),
-                    'Settlement Amount £' => number_format($tx['amount']/100, 2),
+                    'Deposit' =>  number_format( $settlement['deposit']/100,2,'.',''),
```
Comment (srleach) [inline app/Http/Controllers/SettlementsController.php]: CS

File: app/Http/Controllers/SettlementsController.php [hunk 10]
```diff
@@ -127,21 +131,49 @@ private function flattenReport(array $report)
...
+                    'Settlement Amount' => number_format($tx['amount']/100,2,'.',''),
```
Comment (srleach) [inline app/Http/Controllers/SettlementsController.php]: CS

File: app/Http/Controllers/SettlementsController.php [hunk 11]
```diff
@@ -127,21 +131,49 @@ private function flattenReport(array $report)
...
+    {
+        foreach ($report['settlements'] as $settlement) {
+            $rtn[] = [
+                'Order Date' => date('d/m/Y', strtotime($settlement['received_date'])),
+                'Notification Date' => date('d/m/Y', strtotime($settlement['captured_date'])),
+                'Customer' => $settlement['customer_name'],
+                'Post Code' => $settlement['application_postcode'],
+                'Application ID' =>  $settlement['application'],
+                'Retailer Reference' => $settlement['order_reference'],
+                'Order Amount' => number_format($settlement['order_amount']/100,2,'.',''),
```
Comment (srleach) [inline app/Http/Controllers/SettlementsController.php]: CS

File: app/Http/Controllers/SettlementsController.php [hunk 12]
```diff
@@ -127,21 +131,49 @@ private function flattenReport(array $report)
...
+                'Type' => $settlement['type'],
+                'Deposit' => number_format($settlement['deposit']/100,2,'.',''),
```
Comment (srleach) [inline app/Http/Controllers/SettlementsController.php]: CS

File: app/Http/Controllers/SettlementsController.php [hunk 13]
```diff
@@ -127,21 +131,49 @@ private function flattenReport(array $report)
...
+                'Loan Amount' => number_format($settlement['loan_amount']/100,2,'.',''),
```
Comment (srleach) [inline app/Http/Controllers/SettlementsController.php]: CS

File: app/Http/Controllers/SettlementsController.php [hunk 14]
```diff
@@ -127,21 +131,49 @@ private function flattenReport(array $report)
...
+                'Subsidy' => number_format($settlement['subsidy']/100,2,'.',''),
```
Comment (srleach) [inline app/Http/Controllers/SettlementsController.php]: CS

File: app/Http/Controllers/SettlementsController.php [hunk 15]
```diff
@@ -127,21 +131,49 @@ private function flattenReport(array $report)
...
+                'Adjustment' => number_format($settlement['adjustment']/100,2,'.',''),
```
Comment (srleach) [inline app/Http/Controllers/SettlementsController.php]: CS

File: app/Http/Controllers/SettlementsController.php [hunk 16]
```diff
@@ -127,21 +131,49 @@ private function flattenReport(array $report)
...
+                'Settlement Amount' => number_format($settlement['net']/100,2,'.',''),
```
Comment (srleach) [inline app/Http/Controllers/SettlementsController.php]: CS

File: resources/views/settlements/settlement_report.blade.php [hunk 17]
```diff
@@ -18,49 +19,51 @@
...
+                    <tr>
+                        <td>{{ date('d/m/Y', strtotime($item['received_date'])) }}</td>
+                        <td>{{ date('d/m/Y', strtotime($item['captured_date'])) }}</td>
+                        <td>{{ $item['customer_name']}}</td>
+                        <td>{{ $item['application_postcode'] }}</td>
+                        <td>{{ $item['application'] }}</td>
+                        <td>{{ $item['order_reference'] }}</td>
+                        <td class="text-right">{{ '&pound;' . number_format($item['order_amount']/100, 2)}}</td>
+                        <td>{{ $item['type'] }}</td>
+                        <td class="{{($item['deposit'] < 0 ? 'text-danger' : '') }} text-right">{{'&pound;' . number_format($item['deposit']/100,2)}}</td>
```
Comment (srleach) [inline resources/views/settlements/settlement_report.blade.php]: CS x5

File: tests/Controllers/SettlementControllerTest.php [hunk 18]
```diff
@@ -138,23 +138,25 @@ public function setUp()
     /**
      * @author EA
      */
-    public function testFlattenReport()
+    public function testFlattenApiReport()
     {
+
+
         $lastRow =
             [
```
Comment (srleach) [inline tests/Controllers/SettlementControllerTest.php]: CS - Check your alignment here, looks wrong.

**Comment by srleach** [comment]
@00eadeyemi

**Comment by srleach** [comment]
## Scrutiniser

`SettlementsController:156`
`$rtn was never initialized. Although not strictly required by PHP, it is generally a good practice to add $rtn = array(); before regardless.`

File: resources/views/settlements/settlement_report.blade.php [hunk 19]
```diff
@@ -3,9 +3,10 @@
 @section('content')
 
     <h1>
-        SETTLEMENT REPORT
-        <div class="btn-group pull-right">
-            <a href="{!! Request::url() !!}/?download=csv&amp;limit=5000" class="btn btn-default"><span class="glyphicon glyphicon-save" aria-hidden="true"></span> Download CSV</a>
+        Settlement Report
+        <div class="btn-toolbar pull-right">
+            <a href="{{Request::url()}}/?download=csv&amp;source=view_data&amp;filename=export_view_filename&amp;limit=5000" class="btn btn-default"><span class="glyphicon glyphicon-save"></span> Download Report</a>
+            <a href="{!! Request::url() !!}/?download=csv&amp;filename=export_api_filename&amp;limit=5000" class="btn btn-default"><span class="glyphicon glyphicon-save" aria-hidden="true"></span>Download Raw</a>
```
Comment (00eadeyemi) [inline resources/views/settlements/settlement_report.blade.php]: {!! !!} or {{ }} ??? (From Sean)

**Comment by 00eadeyemi** [comment]
@srleach please reassess

File: app/Http/Controllers/SettlementsController.php [hunk 20]
```diff
@@ -107,9 +107,13 @@ public function settlementReport($merchant, $id)
         return View('settlements.settlement_report', [
             'settlementReport' => $settlementReport,
             'installation' => Application::where('ext_id', '=', $settlementReport['id'])->first(),
-            'api_data' => $this->flattenReport($settlementReport),
-            'export_custom_filename' => 'settlement-report-'.$settlementReport['id'].'-'
-                .date_format(DateTime::createFromFormat('Y-m-d',$settlementReport['settlement_date']),'Ymd'),
+            'api_data' => $this->flattenRawReport($settlementReport),
+            'export_api_filename' => 'settlement-raw-' . $settlementReport['id'] . '-'
+                .date_format(DateTime::createFromFormat('Y-m-d', $settlementReport['settlement_date']), 'Ymd'),
```
Comment (srleach) [inline app/Http/Controllers/SettlementsController.php]: CS

File: app/Http/Controllers/SettlementsController.php [hunk 21]
```diff
@@ -107,9 +107,13 @@ public function settlementReport($merchant, $id)
...
+            'view_data' => $this->flattenViewReport($settlementReport),
+            'export_view_filename' => 'settlement-report-' . $settlementReport['id'] . '-'
+                .date_format(DateTime::createFromFormat('Y-m-d', $settlementReport['settlement_date']), 'Ymd'),
```
Comment (srleach) [inline app/Http/Controllers/SettlementsController.php]: CS

File: resources/views/settlements/settlement_report.blade.php [hunk 22]
```diff
@@ -3,9 +3,10 @@
 @section('content')
 
     <h1>
-        SETTLEMENT REPORT
-        <div class="btn-group pull-right">
-            <a href="{!! Request::url() !!}/?download=csv&amp;limit=5000" class="btn btn-default"><span class="glyphicon glyphicon-save" aria-hidden="true"></span> Download CSV</a>
+        Settlement Report
+        <div class="btn-toolbar pull-right">
+            <a href="{{ Request::url() }}/?download=csv&amp;source=view_data&amp;filename={{ $export_view_filename }}" class="btn btn-default"><span class="glyphicon glyphicon-save"></span> Download Report</a>
+            <a href="{{ Request::url() }}/?download=csv&amp;filename={{ $export_api_filename }}" class="btn btn-default"><span class="glyphicon glyphicon-save" aria-hidden="true"></span>Download Raw</a>
```
Comment (srleach) [inline resources/views/settlements/settlement_report.blade.php]: 😉 Think you can find what's wrong?

File: tests/Controllers/SettlementControllerTest.php [hunk 23]
```diff
@@ -199,4 +198,36 @@ public function testApplySettlementAmounts()
...
+            'Type' => "Cancellation",
+            'Deposit' => "-60.00",
+            'Loan Amount' => "-540.00",
+            'Subsidy' => "1.00",
+            'Adjustment' => "0.00",
+            'Settlement Amount' => "-599.00",
+        ];
+
+        $method1 = new ReflectionMethod('App\Http\Controllers\SettlementsController', 'applySettlementAmounts');
+        $method1->setAccessible(TRUE);
```
Comment (srleach) [inline tests/Controllers/SettlementControllerTest.php]: TRUE?

File: tests/Controllers/SettlementControllerTest.php [hunk 24]
```diff
@@ -199,4 +198,36 @@ public function testApplySettlementAmounts()
...
+        $method1->invokeArgs($this->settlementControllerObject, [&$this->settlementReport]);
+
+        $method = new ReflectionMethod('App\Http\Controllers\SettlementsController', 'flattenViewReport');
+        $method->setAccessible(TRUE);
```
Comment (srleach) [inline tests/Controllers/SettlementControllerTest.php]: TRUE? I think you're from cornwall chap 😉

Comment (srleach) [inline tests/Controllers/SettlementControllerTest.php]: TRUE? I think you mean true

File: tests/Controllers/SettlementControllerTest.php [hunk 25]
```diff
@@ -138,27 +138,26 @@ public function setUp()
...
+            'Notification Date' => '06/06/2016',
+            'Type' => 'Manual Adjustment',
+            'Description' => 'Order from TestCheckout',
+            'Deposit' => '60.00',
+            'Settlement Amount'=> '-20.00',
+        ];
 
-        $method = new ReflectionMethod('App\Http\Controllers\SettlementsController', 'flattenReport');
+        $method = new ReflectionMethod('App\Http\Controllers\SettlementsController', 'flattenRawReport');
         $method->setAccessible(TRUE);
```
Comment (srleach) [inline tests/Controllers/SettlementControllerTest.php]: TRUE?

**Comment by srleach** [comment]
@00eadeyemi Back to you.

File: tests/Controllers/SettlementControllerTest.php [hunk 26]
```diff
@@ -199,4 +198,36 @@ public function testApplySettlementAmounts()
...
 
     }
+
+    /**
+     * @author EA
+     */
+    public function testFlattenViewReport()
+    {
+        $lastRow = [
+            'Order Date' => "31/05/2016",
```
Comment (srleach) [inline tests/Controllers/SettlementControllerTest.php]: CS* 12

File: app/Http/Controllers/SettlementsController.php (hunk 20 above)
Comment (srleach) [inline app/Http/Controllers/SettlementsController.php]: This is still here, third time now...

File: tests/Controllers/SettlementControllerTest.php (hunk 25 above)
Comment (srleach) [inline tests/Controllers/SettlementControllerTest.php]: ...?

**Comment by 00eadeyemi** [comment]
@srleach please reassess

**Comment by srleach** [comment]
@00eadeyemi Subject to scrutiniser passing this looks ok. @wnowicki

**Comment by wnowicki** [comment]
- OK
- Waiting for CI

**Pull request merged**